lib/kube/proxy/auth.go

```diff
@@ -53,83 +53,107 @@ func TestOnlySkipSelfPermissionCheck(skip bool) {
+// - pod service account credentials: files in hardcoded paths when running
+//   inside of a k8s pod; this is used when kubeClusterName is set
+// - kubeconfig: a file with a set of k8s endpoints and credentials mapped to
+//   theml this is used when kubeconfigPath is set
+//
+// newKubeService changes the loading behavior:
+// - false:
+//   - if loading from kubeconfig, only "current-context" is returned; the
+//     returned map key matches tpClusterName
+//   - if no credentials are loaded, no error is returned
+//   - permission self-test failures are only logged
+// - true:
+//   - if loading from kubeconfig, all contexts are returned
+//   - if no credentials are loaded, returns an error
+//   - permission self-test failures cause an error to be returned
+func getKubeCreds(ctx context.Context, log logrus.FieldLogger, tpClusterName, kubeClusterName, kubeconfigPath string, newKubeService bool) (map[string]*kubeCreds, error) {
+	log.
+		WithField("kubeconfigPath", kubeconfigPath).
+		WithField("kubeClusterName", kubeClusterName).
+		WithField("newKubeService", newKubeService).
+		Debug("Reading kubernetes creds.")
+
+	// Load kubeconfig or local pod credentials.
+	cfg, err := kubeutils.GetKubeConfig(kubeconfigPath, newKubeService, kubeClusterName)
+	if err != nil && !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+	if trace.IsNotFound(err) || len(cfg.Contexts) == 0 {
+		if newKubeService {
+			return nil, trace.BadParameter("no kubernetes credentials found; kubernetes_service requires either a valid kubeconfig_path or to run inside of a kubernetes pod")
 		}
+		log.Debugf("Could not load kubernetes credentials. This proxy will still handle kubernetes requests for trusted teleport clusters or kubernetes nodes in this teleport cluster")
+		return nil, nil
+	}
+	if !newKubeService {
+		// Hack for proxy_service - register a k8s cluster named after the
+		// teleport cluster name to route legacy requests.
+		//
+		// Also, remove all other contexts. Multiple kubeconfig entries are
+		// only supported for kubernetes_service.
+		cfg.Contexts = map[string]*rest.Config{
+			tpClusterName: cfg.Contexts[cfg.CurrentContext],
+		}
+	}
 
-		cfg, err = kubeutils.GetKubeConfig(os.Getenv(teleport.EnvKubeConfig))
+	res := make(map[string]*kubeCreds, len(cfg.Contexts))
+	// Convert kubeconfig contexts into kubeCreds.
+	for cluster, clientCfg := range cfg.Contexts {
+		log := log.WithField("cluster", cluster)
+		log.Debug("Checking kubernetes impersonation permissions.")
+		client, err := kubernetes.NewForConfig(clientCfg)
 		if err != nil {
-			return nil, trace.BadParameter(`auth server assumed that it is
-running in a kubernetes cluster, but could not init in-cluster kubernetes client: %v`, err)
+			return nil, trace.Wrap(err, "failed to generate kubernetes client for cluster %q: %v", cluster, err)
+		}
+		// For each loaded cluster, check impersonation permissions. This
+		// should only fail for newKubeService.
```

Comment:
Nit: "This should only fail for newKubeService" gives a wrong impression, I would rephrase to "The failure is only critical error for a newKubeService".

Reply:
done